src/libcollections/slice.rs

```diff
@@ -1305,213 +1304,290 @@ impl<T: Clone> ToOwned for [T] {
-    // manipulate/step the pointers themselves, rather than repeatedly
-    // .offset-ing.
-    for start in (0..len).step_by(insertion) {
-        // start <= i < len;
-        for i in start..cmp::min(start + insertion, len) {
-            // j satisfies: start <= j <= i;
-            let mut j = i as isize;
-            unsafe {
-                // `i` is in bounds.
-                let read_ptr = buf_v.offset(i as isize);
+    // When dropped, copies the range `start..end` into `dst`.
+    struct Guard<T> {
+        start: *mut T,
+        end: *mut T,
+        dst: *mut T,
+    }
 
-                // find where to insert, we need to do strict <,
-                // rather than <=, to maintain stability.
+    impl<T> Drop for Guard<T> {
+        fn drop(&mut self) {
+            let len = (self.end as usize - self.start as usize) / mem::size_of::<T>();
+            unsafe { ptr::copy_nonoverlapping(self.start, self.dst, len); }
+        }
+    }
+}
 
-                // start <= j - 1 < len, so .offset(j - 1) is in
-                // bounds.
-                while j > start as isize && compare(&*read_ptr, &*buf_dat.offset(j - 1)) == Less {
-                    j -= 1;
-                }
+/// This merge sort borrows some (but not all) ideas from TimSort, which is described in detail
+/// [here](http://svn.python.org/projects/python/trunk/Objects/listsort.txt).
+///
+/// The algorithm identifies strictly descending and non-descending subsequences, which are called
+/// natural runs. There is a stack of pending runs yet to be merged. Each newly found run is pushed
+/// onto the stack, and then some pairs of adjacent runs are merged until these two invariants are
+/// satisfied, for every `i` in `0 .. runs.len() - 2`:
+///
+/// 1. `runs[i].len > runs[i + 1].len`
+/// 2. `runs[i].len > runs[i + 1].len + runs[i + 2].len`
+///
+/// The invariants ensure that the total running time is `O(n log n)` worst-case.
+fn merge_sort<T, F>(v: &mut [T], mut compare: F)
+    where F: FnMut(&T, &T) -> Ordering
+{
+    // When arrays have length up to this value, insertion sort is going to perform better than
+    // merge sort. For bigger types `T`, the threshold is a bit smaller.
+    // FIXME #12092: These numbers are platform-specific and need more extensive testing/tuning.
+    let insertion = if size_of::<T>() <= 16 { 32 } else { 16 };
 
-                // shift everything to the right, to make space to
-                // insert this value.
+    let len = v.len();
 
-                // j + 1 could be `len` (for the last `i`), but in
-                // that case, `i == j` so we don't copy. The
-                // `.offset(j)` is always in bounds.
-                ptr::copy(&*buf_dat.offset(j), buf_dat.offset(j + 1), i - j as usize);
-                ptr::copy_nonoverlapping(read_ptr, buf_dat.offset(j), 1);
+    // Short arrays get sorted in-place via insertion sort to avoid allocations.
+    if len <= insertion {
```

Comment:
It occurs to me that this threshold might want to be higher than the threshold for sorting runs, since it might be worth some time penalty to avoid the allocation.